StartBonusActivation/handler.ts

```diff
@@ -1,51 +1,283 @@
+        >
+      >(fromEither(left(ResponseErrorInternal(`Cannot find DSU`))), doc =>
+        // the found document is not in eligible status
+        !EligibilityCheckSuccessEligible.is(doc)
+          ? fromEither(left(ResponseErrorForbiddenNotAuthorized))
+          : // the check is expired
+          doc.validBefore > new Date()
+          ? fromEither(left(ResponseErrorGone(`DSU expired`)))
+          : // the check is fine, I can extract the DSU data from it
+            fromEither(
+              right(
+                // extract dsu keys from EligibilityCheckSuccessEligible
+                pick(keys(Dsu._A), doc)
+              )
+            )
+      );
+    }
+  );
+
+/**
+ * Create a new BonusActivation request record
+ * @param bonusActivationModel an instance of BonusActivationModel
+ * @param fiscalCode the id of the requesting user
+ * @param dsu the valid DSU of the current user
+ *
+ * @returns either the created record or
+ */
+const createBonusActivation = (
+  bonusActivationModel: BonusActivationModel,
+  fiscalCode: FiscalCode,
+  dsu: Dsu
+): TaskEither<IResponseErrorInternal, BonusActivation> =>
+  fromQueryEither(() => {
+    const bonusCode = "any code" as NonEmptyString;
+    const bonusActivation: NewBonusActivation = {
+      applicantFiscalCode: fiscalCode,
+      code: bonusCode,
+      dsuRequest: dsu,
+      id: bonusCode,
+      kind: "INewBonusActivation",
+      status: BonusActivationStatusEnum.PROCESSING,
+      updatedAt: new Date()
+    };
+    return bonusActivationModel.create(bonusActivation, fiscalCode);
+  }).mapLeft(err =>
+    ResponseErrorInternal(`Error creating BonusActivation: ${err.message}`)
+  );
+
+type StartBonusActivationResponse =
+  | IResponseErrorInternal
+  | IResponseErrorForbiddenNotAuthorized
+  | IResponseErrorGone
+  | IResponseSuccessAccepted
+  | IResponseSuccessRedirectToResource<BonusActivation, BonusActivation>;
 
 type IStartBonusActivationHandler = (
   context: Context,
   fiscalCode: FiscalCode
-) => Promise<
-  // tslint:disable-next-line: max-union-size
-  | IResponseSuccessAccepted
-  // TODO: Add types
-  | IResponseSuccessRedirectToResource<unknown, unknown>
-  | IResponseErrorInternal
-  | IResponseErrorConflict
->;
+) => Promise<StartBonusActivationResponse>;
 
-export function StartBonusActivationHandler(): IStartBonusActivationHandler {
+export function StartBonusActivationHandler(
+  bonusActivationModel: BonusActivationModel,
+  eligibilityCheckModel: EligibilityCheckModel
+): IStartBonusActivationHandler {
   return async (context, fiscalCode) => {
-    // TODO: Add implementation
     const client = df.getClient(context);
-    const status = await client.getStatus(
-      makeStartBonusActivationOrchestratorId(fiscalCode)
-    );
-    if (status.runtimeStatus === df.OrchestrationRuntimeStatus.Running) {
-      return ResponseSuccessAccepted();
-    }
-    return ResponseErrorConflict("Implementation missing");
+
+    // TODO: lock for familiuid
```

Comment:
sorry I meant: these TODOs sould be put wherever the implementation goes (not as a checklist before everything else : )

Reply:
done